src/main/java/com/solbeg/Main.java

```diff
@@ -129,6 +144,8 @@ public static Optional<User> retrieveUserGmail(UserProvider userProvider) {
      */
     public static User getUserWithFallback(UserProvider userProvider, UserProvider fallbackProvider) {
 
+        return userProvider.getUser().orElseGet(() -> fallbackProvider.getUser()
+                .orElseThrow(NoSuchElementException::new));
```

Comment:
this will work fine but you can improve it using static reference with or() method
![image](https://github.com/MusienkoYuriy97/optional-api-advanced/assets/62200771/adca5904-a621-4e43-a6fb-f22906616918)

Reply:
fixed